Documentation enhancements and add doxygen references

nashif opened the pull request:
- Move data structure into their own folder and toc
- Use :c:func: where possible
- Merge ring buffer docs
- Move k_work API docs to work queue documentation (was in threads)
- add missing doxygen grouping in multiple places.

nashif committed:
kernel: move workqueue API reference

Move workqueue doxygen docs to its own group and reference that in RST
under the workqueue section.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: group data structure docs and add doxygen linkage

Move out of misc/ and put in own folder and add the grouping to doxygen
to be able to reference the doxygen docs into RST.
Move each item into their own file to reduce clutter and to make it
less crowded in one single page.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
docs: structures: move ring buffer

We had duplicate entries for ring buffer, merge them into one document.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: strcutures: use doxygen to reference functions

Use :c:func:`..` possible to reference and link doxygen documentation.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

gmarull commented on doc/reference/data_structures/dlist.rst:
```diff
@@ -12,38 +12,38 @@ pointers per node, and thus has somewhat higher runtime code and
-accessible memory.  It must be initialized with ``sys_dlist_init()``
-or ``SYS_DLIST_STATIC_INIT()`` before use.  The ``sys_dnode_t`` struct
+accessible memory.  It must be initialized with :c:func:`sys_dlist_init`
+or :c:macro:`SYS_DLIST_STATIC_INIT` before use.  The ``sys_dnode_t`` struct
```
do `c:type:` or `c:struct:` work for stuff like `sys_dnote_t` (and others)?

nashif replied:
they could, if they are available through doxygen though.

By the way, I noticed that we do not get a failure when we use :c:func:`...` on something that is not defined in doxygen, is that normal?

gmarull replied:
I don't know, that could be the way Sphinx C domain handles missing references (maybe https://github.com/sphinx-doc/sphinx/blob/4.x/sphinx/domains/c.py#L3840 ?).

nashif committed:
doc: enable doxygen for CONFIG_TIMING_FUNCTIONS

This was missed and APIs were not showing in final docs.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: modbus: remove extra level in the docs

We had one extra level in the docs, reduce and show modbus details
directly instead of the extra step.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: rename Random -> RNG

Random alone in the TOC is misleading and confusing, so be more
specific.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: smp: use doxygen references

Use :c:func:`..`` where possible to reference documented APIs

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
